pkg/sample/random_test.go

```diff
@@ -0,0 +1,867 @@
+				if tc.a != 0 && tc.b != 0 && len(x) != len(y) {
+					t.Fatalf("expected %#v got %#v", len(x), len(y))
+				}
+
+				{
+					if generic.Duplicate(x) {
+						t.Fatalf("expected x to not contain duplicates")
+					}
+
+					if generic.Duplicate(y) {
+						t.Fatalf("expected y to not contain duplicates")
+					}
+				}
+
+				// Below we allocate slices to access items using the generated indices.
+				// If we are able to access those items using any given index without
+				// creating a runtime panic, then our implementation is safe to the
+				// extend that no indices are generated out of bounds.
+				{
+					slx := make([]int, tc.a)
+					for _, i := range x {
+						_ = slx[i]
+					}
+
+					sly := make([]int, tc.b)
+					for _, i := range y {
+						_ = sly[i]
+					}
+				}
+			}
+
+			// We somehow want to assert that we generate a variation of random
+			// indices. Below we check that there are some minimum amounts of
+			// generated permutations, which implies that each of the existing keys in
+			// the associated map of observed indices was generated at least once.
+			// More keys equals more randomness. And so we compare against the
+			// smallest common denominator of the given test case.
+			{
+				lnx := uint(len(mpx))
+				mnx := min(tc.a, tc.b)
+				if lnx < mnx {
+					t.Fatalf("expected x to generate at least %d different variations, got %d", lnx, mnx)
+				}
+
+				lny := uint(len(mpy))
+				mny := min(tc.a, tc.b)
+				if lny < mny {
+					t.Fatalf("expected y to generate at least %d different variations, got %d", lny, mny)
+				}
+			}
+		})
+	}
+}
+
+func Test_Sample_Random_Fake(t *testing.T) {
+	testCases := []struct {
+		a uint
+		b uint
+		s int
+		x []int
+		y []int
+	}{
+		// Case 000
+		{
+			a: 0,
+			b: 0,
+			s: 0,
+			x: nil,
+			y: nil,
+		},
+		// Case 001
+		{
+			a: 0,
+			b: 0,
+			s: 3,
+			x: nil,
+			y: nil,
+		},
+
+		// Case 002
+		{
+			a: 1,
+			b: 0,
+			s: 0,
+			x: []int{
+				0,
+			},
+			y: nil,
+		},
+		// Case 003
+		{
+			a: 0,
+			b: 1,
+			s: 0,
+			x: nil,
+			y: []int{
+				0,
+			},
+		},
+
+		// Case 004
+		{
+			a: 1,
+			b: 0,
+			s: 2376,
+			x: []int{
+				0,
+			},
+			y: nil,
+		},
+		// Case 005
+		{
+			a: 0,
+			b: 1,
+			s: 2376,
+			x: nil,
+			y: []int{
+				0,
+			},
+		},
+
+		// Case 006
+		{
+			a: 3,
+			b: 0,
+			s: 0,
+			x: []int{
+				2,
+			},
+			y: nil,
+		},
+		// Case 007
+		{
+			a: 0,
+			b: 3,
+			s: 0,
+			x: nil,
+			y: []int{
+				2,
+			},
+		},
+
+		// Case 008
+		{
+			a: 3,
+			b: 0,
+			s: 665117,
+			x: []int{
+				1,
+			},
+			y: nil,
+		},
+		// Case 009
+		{
+			a: 0,
+			b: 3,
+			s: 665117,
+			x: nil,
+			y: []int{
+				1,
+			},
+		},
+
+		// Case 010
+		{
+			a: 10,
+			b: 0,
+			s: 0,
+			x: []int{
+				9,
+			},
+			y: nil,
+		},
+		// Case 011
+		{
+			a: 0,
+			b: 10,
+			s: 0,
+			x: nil,
+			y: []int{
+				9,
+			},
+		},
+
+		// Case 012
+		{
+			a: 10,
+			b: 0,
+			s: 104017,
+			x: []int{
+				6,
+			},
+			y: nil,
+		},
```

Comment:
At first glance this test might be confusing. If there are 10 stakers we should select 2 users. Here there are 10 users on one side, while there are no users on the other. And so the minimum we can draw is only 1 random address based on the current implementation.